### ISSUE TITLE
Add support for stopping currently running tests

### ISSUE THREAD

**Issue by anadahz** [ooni/probe#590]
Accidentally I hit run to the same test so now I have two tests running concurrently with no way to stop any of them.

Perhaps it makes sense to add a way for a user to stop or force stop a current running test.

**Comment by hellais** [comment]
I looked into how much effort it would be to implement this and it seems like it's actually not something that is easily achievable.

I would say this is more of a feature request than a bug report and I would put it inside of the backlog.

**Comment by darkk** [comment]
Also, it is possible to have two instances of same decks running concurrently: one started manually, another launched by ooniprobe-agent scheduler.

**Comment by hellais** [comment]
I believe this would require some features inside of measurement-kit. @bassosimone do you have ideas on how we could/would do this?
I there already a ticket on the mk side to support this?

**Comment by bassosimone** [comment]
The OP is related to concurrently running tests, which cannot happen in MK.

The possibility of interrupting tests was proposed as a solution for the OP.

I have started adding support for interrupting tests on MK. A tracking ticker should probably be added, tho.

Still: is this a feature we want? Interrupting tests adds code complexity. I remember @darkk wisely saying “when I want to interrupt a test, I kill the app”.

Another possible approach to solve the OP problem is either queuing tests (more or less the current approach) or preventing further tests from running.

@hellais @darkk do you think we really need to make tests interruptible or is that optional complexity that we can avoid adding this allowing our code to be simpler?

I used to have a preference for allowing to interrupt but since then I changed my mind and think this is extra complexity.

**Comment by hellais** [comment]
I think it would be useful to have some logic that allows a user to do a "clean interruption" of a currently running test.

The current execution model of the desktop and mobile apps is that the UI is "blocking" when you are running a test and there is no way to interrupt them. I think this is fine for the moment, but I don't think it's optimal.

If users do "kill the app" to stop a test, this means that we have unclosed reports in the collector that need to be spooled differently.
Ideally there would be a stop button that would work by:
1. If it's a multi-entry test it would run the test until the end of the current input and then go the shutdown logic (i.e. close the report and do the finalise stage)
2. If it's not a multi-entry test, I think it's fine if it just finishes the currently running test until the end or if the user does a hard stop (equivalent of pressing ctrl-c twice on the terminal), then it will interrupt the current running test and do the shutdown phase. In this case we should probably have something in the collector to mark a currently open report as invalid.

So basically from the MK side of things it would be useful to have something to signal that:

a. The execution should stop once the next input is processed (similar to the max_runtime option)
b. The execution should stop immediately and signal the collector that the current report that was opened is actually invalid

Is this what you were also thinking of or is this simpler than what you had in mind?

**Comment by bassosimone** [comment]
(Terse rely because from mobile) I was writing the equivalent of a golang Context object where there is the interrupted flag and was planning to passing it down to tests and to check it to decide whether to interrupt. I agree with what you say wrt stopping a test with input vis a vis stopping a test without input.

**Comment by hellais** [comment]
This has been implemented in the most recent version of probe-desktop.

It allows you to stop a test, but you can't actually see the results because the test is not marked as done.

For the next steps we have to do for this I filed: https://github.com/ooni/probe-cli/issues/76
